lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureMultipartOutputStream.java

```diff
@@ -0,0 +1,248 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.azure;
+
+import com.azure.storage.blob.BlobClient;
+import com.azure.storage.blob.specialized.BlobOutputStream;
+import com.azure.storage.blob.specialized.BlockBlobClient;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.trino.memory.context.AggregatedMemoryContext;
+import io.trino.memory.context.LocalMemoryContext;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.util.concurrent.Futures.submit;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static io.trino.filesystem.azure.AzureUtils.handleAzureException;
+import static java.lang.Math.clamp;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.System.arraycopy;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class AzureMultipartOutputStream
+        extends OutputStream
+{
+    private final AzureLocation location;
+    private final BlockBlobClient blockClient;
+    private final LocalMemoryContext memoryContext;
+    private final Executor uploadExecutor;
+
+    private final int writeBlockSizeBytes;
+    private final boolean overwrite;
+    private boolean closed;
+    private boolean multiStagingStarted;
+    private final List<String> stagedBlocks = new ArrayList<>();
+    private final List<ListenableFuture<?>> stagedBlockFutures = new ArrayList<>();
+    private int blockNum;
+
+    private byte[] buffer = new byte[0];
+    private int bufferSize;
+    private int initialBufferSize = 64;
+
+    public AzureMultipartOutputStream(
+            AzureLocation location,
+            BlobClient blobClient,
+            ExecutorService uploadExecutor,
+            boolean overwrite,
+            AggregatedMemoryContext memoryContext,
+            int writeBlockSizeBytes)
+    {
+        requireNonNull(location, "location is null");
+        requireNonNull(blobClient, "blobClient is null");
+        checkArgument(writeBlockSizeBytes >= 0, "writeBlockSizeBytes is negative");
+
+        this.location = location;
+        this.writeBlockSizeBytes = writeBlockSizeBytes;
```

Comment:
`s3.streaming.part-size` default is 32MB. We've seen low values cause throttling in S3 (https://github.com/trinodb/trino/pull/25781).
Is the 4MB default here too low ?